espnet/nets/batch_beam_search_online.py

```diff
@@ -0,0 +1,279 @@
+            ):
+                logging.info(f"end detected at {self.process_idx}")
+                return self.assemble_hyps(self.ended_hyps)
+
+            if len(local_ended_hyps) > 0 and not is_final:
+                logging.info("Detected hyp(s) reaching EOS in this block.")
+                break
+
+            self.prev_hyps = self.running_hyps
+            self.running_hyps = self.post_process(
+                self.process_idx, maxlen, maxlenratio, best, self.ended_hyps
+            )
+
+            if is_final:
+                for hyp in local_ended_hyps:
+                    self.ended_hyps.append(hyp)
+
+            if len(self.running_hyps) == 0:
+                logging.info("no hypothesis. Finish decoding.")
+                return self.assemble_hyps(self.ended_hyps)
+            else:
+                logging.debug(f"remained hypotheses: {len(self.running_hyps)}")
+            # increment number
+            self.process_idx += 1
+
+        if is_final:
+            return self.assemble_hyps(self.ended_hyps)
+        else:
+            for hyp in self.ended_hyps:
+                local_ended_hyps.append(hyp)
+            rets = self.assemble_hyps(local_ended_hyps)
+
+            if self.process_idx > 1 and len(self.prev_hyps) > 0:
+                self.running_hyps = self.prev_hyps
+                self.process_idx -= 1
+                self.prev_hyps = []
+
+            return rets
+
+        
+    def assemble_hyps(self, ended_hyps):
+        nbest_hyps = sorted(ended_hyps, key=lambda x: x.score, reverse=True)
+        # check the number of hypotheses reaching to eos
+        if len(nbest_hyps) == 0:
+            logging.warning(
+                "there is no N-best results, perform recognition "
+                "again with smaller minlenratio."
+            )
+            return (
+                []
+            )
+
+        # report the best result
+        best = nbest_hyps[0]
+        for k, v in best.scores.items():
+            logging.info(
+                f"{v:6.2f} * {self.weights[k]:3} = {v * self.weights[k]:6.2f} for {k}"
+            )
+        logging.info(f"total log probability: {best.score:.2f}")
+        logging.info(f"normalized log probability: {best.score / len(best.yseq):.2f}")
+        logging.info(f"total number of ended hypotheses: {len(nbest_hyps)}")
+        if self.token_list is not None:
+            logging.info(
+                "best hypo: "
+                + "".join([self.token_list[x] for x in best.yseq[1:-1]])
+                + "\n"
+            )
+        return nbest_hyps
+        
+    def extend(self, x: torch.Tensor, hyps: Hypothesis) -> List[Hypothesis]:
+        """Extend probabilities and states with more encoded chunks.
+
+        Args:
+            x (torch.Tensor): The extended encoder output feature
+            hyps (Hypothesis): Current list of hypothesis
+
+        Returns:
+            Hypothesis: The exxtended hypothesis
```

Comment:
typo: extended.

btw can you briefly summarize what kind of methods are required (e.g., extend_prob, extend_state) in the class-level (or module-level) docstring?